sql: CREATE INDEX inherits PARTITION BY on PARTITION ALL BY tables

otan opened the pull request:
Resolves #58734 

Release note (sql change): CREATE INDEX will now inherit PARTITION BY
clauses from PARTITION ALL BY tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58988)
<!-- Reviewable:end -->

ajstorm reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58988)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/create_index.go, line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58988#-MR08bqMBxbh4pctC5cM:-MR08bqN9TD6YNUnGnXh:b851et7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc87c5482d6731bbf8ebfe300a7eb3b53229703f/pkg/sql/create_index.go#L437)):*
> ```Go
> 		a := &rowenc.DatumAlloc{}
> 		f := tree.NewFmtCtx(tree.FmtSimple)
> 		if err := ShowCreatePartitioning(
> ```

It seems strange to me that we need to build the SHOW statement and parse it to derive the PartitionBy info.  Are there other options at our disposal here?  Could we build this using the table's PartitionDescriptor?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/create_index.go, line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58988#-MR08bqMBxbh4pctC5cM:-MR1fpRb0YIdnEw1L_kh:b-16g477) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc87c5482d6731bbf8ebfe300a7eb3b53229703f/pkg/sql/create_index.go#L437)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

It seems strange to me that we need to build the SHOW statement and parse it to derive the PartitionBy info.  Are there other options at our disposal here?  Could we build this using the table's PartitionDescriptor?
</blockquote></details>

the logic to go from the persisted partitioning descriptor back to the `tree.PartitionBy` is actually fairly complex.
we do this kind of logic for type parsing and the like, and it seems easier than doing the same thing twice ;) 
if you want i can leave a TODO to not rely on it, or we can defer this to later. for MVP purposes though i think this does the job in by far the most minimal lines of code :P 


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58988)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/create_index.go, line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58988#-MR08bqMBxbh4pctC5cM:-MR1rGhKCdoFCP7KPwKk:b-k9ulhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc87c5482d6731bbf8ebfe300a7eb3b53229703f/pkg/sql/create_index.go#L437)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

the logic to go from the persisted partitioning descriptor back to the `tree.PartitionBy` is actually fairly complex.
we do this kind of logic for type parsing and the like, and it seems easier than doing the same thing twice ;) 
if you want i can leave a TODO to not rely on it, or we can defer this to later. for MVP purposes though i think this does the job in by far the most minimal lines of code :P 
</blockquote></details>

I'm good with a TODO.


<!-- Sent from Reviewable.io -->

otan committed:
sql: CREATE INDEX inherits PARTITION BY on PARTITION ALL BY tables

Release note (sql change): CREATE INDEX will now inherit PARTITION BY
clauses from PARTITION ALL BY tables.

otan commented:
thanks

bors r=ajstorm
